### PR TITLE
Allow <keyboard>.h to be optional when going data driven

### PR DESCRIPTION
**Pull request by zvecr**
## Description

Bit of a long winded change, however the stub header is to avoid..
1. https://kbfirmware.com/ exports using kb.h
2. similar to above, badly named files where `folder != keyboard.h` name

If the above are not important scenarios, then setting `QMK_KEYBOARD_H` to just `quantum.h` is much simpler. However the downside leads to more scenarios where users do the wrong thing, randomly including files here and there till it works.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by zvecr** [comment]
API gen needs fixing so moved back to draft

**Review by Erovia — APPROVED** [review]

**Review by skullydazed — COMMENTED** [review]
Overall this is a great approach. Very glad to see more progress on data driven. I do have one suggestion and one section I'm trying to understand the need for.

File: lib/python/qmk/cli/generate/keyboard_h.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+    has_layout_h = would_populate_layout_h(cli.config.generate_keyboard_h.keyboard)
+
+    # Build the layouts.h file.
+    keyboard_h_lines = ['/* This file was generated by `qmk generate-keyboard-h`. Do not edit or copy.' ' */', '', '#pragma once', '#include "quantum.h"']
+
+    if has_layout_h:
+        keyboard_h_lines.append('#include "layouts.h"')
+    else:
+        keyboard_h_lines.append('#pragma error("<keyboard>.h is only optional for data driven keyboards - kb.h == bad times")')
```
[the remark this answers is not in the document]
Comment (zvecr) [reply lib/python/qmk/cli/generate/keyboard_h.py:57]: Ideally, everything shouldnt include `layouts.h`. However removing that causes some issues right now, and would be better as a 2nd pass.

The build error error is per the PR description.

File: lib/python/qmk/cli/generate/keyboard_h.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    """
+    # Determine our keyboard(s)
+    if not cli.config.generate_keyboard_h.keyboard:
+        cli.log.error('Missing parameter: --keyboard')
+        cli.subcommands['info'].print_help()
+        return False
+
+    if not is_keyboard(cli.config.generate_keyboard_h.keyboard):
+        cli.log.error('Invalid keyboard: "%s"', cli.config.generate_keyboard_h.keyboard)
+        return False
```
[the remark this answers is not in the document]
Comment (zvecr) [reply lib/python/qmk/cli/generate/keyboard_h.py:47]: This is copy pasted from the other 2 generate commands, my suggestion would be to tidy that up.

File: lib/python/qmk/cli/generate/keyboard_h.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (zvecr) [reply lib/python/qmk/cli/generate/keyboard_h.py]: Removed for now, will tackle the layouts.h stuff another time maybe.

**Review by skullydazed — APPROVED** [review]

File: lib/python/qmk/cli/generate/keyboard_h.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (skullydazed) [reply lib/python/qmk/cli/generate/keyboard_h.py]: > Ideally, everything shouldnt include layouts.h. However removing that causes some issues right now, and would be better as a 2nd pass.

Agreed there, but we have more infrastructure to build before that will be possible.

**Pull request merged**